net/socket/socket.c

```diff
@@ -104,9 +104,11 @@ int psock_socket(int domain, int type, int protocol,
 
   /* When usrsock daemon returns -ENOSYS or -ENOTSUP, it means to use
    * kernel's network stack, so fallback to kernel socket.
+   * When -ENETDOWN is returned, it means the usrsock daemon was never
+   * launched or is no longer running, so fallback to kernel socket.
    */
 
-  if (ret == 0 || (ret != -ENOSYS && ret != -ENOTSUP))
+  if (ret == 0 || (ret != -ENOSYS && ret != -ENOTSUP && ret != -ENETDOWN))
```

Comment:
since -ENETDOWN return in other case, should we change the errror code to -ENOSYS/-ENOTSUP if the server doesn't launch yet?

Reply:
As far as I can see, -ENETDOWN is only returned from usrsock_sockif_setup() if either:
a. the application side doesn't have the dev node open. In which case usrsockdev_is_opened() returns false in usrsock_request().
  usrsock_sockif_setup()->usrsock_socket()->do_socket_request()->usrsock_do_request()->usrsock_request returns() -> returns -ENETDOWN

b. The USRSOCK_EVENT_ABORT event is sent to the callback "socket_event" registered in usrsock_socket().
The USRSOCK_EVENT_ABORT is only set in usrsock_abort() which is called when the application closes the /dev/usrsock devnode. Both cases should fallback to kernel provided sockets.


Is there another case of -ENETDOWN that I did not see?

Reply:
Ok, but usrsock test fail with this patch, please fix it.

Reply:
Thanks. I has missed that.

There was a test which aborted the daemon and was checking to see ENETDOWN.
I propagated the original return code of usrsock if it tired to fallback to kernel but the kernel doesn't have any sockets of that type compiled in. Which is what the behaviour would have been before #9647 when it used to only use usrsock if the kernel didnt provide a socket of that family.

The other option was to keep the shutdown test but remove the check for ENETDOWN, but that felt like a reduciton of testing scope.